### PR TITLE
Warn unicode escapes

### PR DESCRIPTION
**Pull request by janschulz**
Warn when a usere inputs a unicode char which would be escaped before execution...

### CODE REVIEW

**Comment by flying-sheep** [comment]
Super nice, especially the wiki page!

https://github.com/IRkernel/repr/wiki/Problems-with-unicode-on-windows

File: R/execution.r [hunk 1]
```diff
@@ -133,6 +134,14 @@ execute = function(request) {
     assign('quit', quit, envir = .GlobalEnv)
     assign('q',    quit, envir = .GlobalEnv)
     
+    send_error_msg <- function(msg){
```
Comment (flying-sheep) [inline R/execution.r]: can you put this as method into the executor (and add a space here)?

maybe like `send_error_msg(request, msg)`? `if (!silent)` would have o go outside of this then though, …

i prefer not to create too many functions in procedural pieces of code.

File: R/execution.r [hunk 2]
```diff
@@ -241,6 +246,23 @@ execute = function(request) {
     interrupted <<- FALSE
     last_recorded_plot <<- NULL
     
+    # Workaround to warn user when code contains potential problematic code
+    # https://github.com/IRkernel/repr/issues/28#issuecomment-208810772
+    # See https://github.com/hadley/evaluate/issues/66
+    if(.Platform$OS.type == "windows") {
```
Comment (flying-sheep) [inline R/execution.r]: " → '

(more below)

File: R/execution.r [hunk 3]
```diff
@@ -241,6 +246,23 @@ execute = function(request) {
...
+        # strip whitespace, because trailing newlines would trip the test...
+        code <- gsub("^\\s+|\\s+$", "", request$content$code)
+        real_len <- nchar(code)
+        r_len <- nchar(paste(capture.output(cat(code)), collapse="\n"))
```
Comment (flying-sheep) [inline R/execution.r]: " → ' again and spaces (also below)

File: R/execution.r [hunk 4]
```diff
@@ -241,6 +246,23 @@ execute = function(request) {
...
+        if (real_len != r_len){
+            msg = c("Your code contains a unicode char which cannot be displayed in your ",
```
Comment (flying-sheep) [inline R/execution.r]: maybe extract to top level as sth. like `win_unicode_warning` and just do

```r
win_unicode_warning <- "Blah
blah blah
blah"
```

File: R/execution.r (hunk 1 above)
Comment (janschulz) [inline R/execution.r]: I don't think it makes sense to put that anywhere else: if the `silent` is outside the function, it's simple a `send_response` call and then I could simple use that...

File: R/execution.r (hunk 2 above)
Comment (janschulz) [inline R/execution.r]: damn it... I really hate Rs style on quotes...

File: R/execution.r (hunk 3 above)
Comment (janschulz) [inline R/execution.r]: ... and spaces... who puts spaces around `=` in function calls?

**Comment by janschulz** [comment]
Will fix code style...

Re code reorg and promoting some to toplevel: I'm not sure how much time is used to redeclare the functions on each execution, but there is a lot of function declaration going on in the execution of code. If it makes a difference, there are a lot of functions which could be declared somewhere else (e.g. as `something(code, silent)` instead of using silent from the outside scope...) or by using a `context` object which the handler could request form an environment?

File: R/execution.r (hunk 3 above)
Comment (flying-sheep) [inline R/execution.r]: certainly not python :laughing:

but yeah, in R that’s how you do it.

File: R/execution.r (hunk 1 above)
Comment (flying-sheep) [inline R/execution.r]: It’s a send_response call with most fields filled in. I always create small helpers like that to reduce complexity

**Comment by flying-sheep** [comment]
I don’t think it makes a difference, it’s just horrible spahghetti code (no offense, i added my share of spaghetti on top)

i think we should extract all declarations and put them extra. that can easily be done by adding all variables they grab from scope (including the executor itself) as parameters.

but here maybe you can simply do `if (!silent) send_error_msg(response, msg)`

**Comment by janschulz** [comment]
As far as I understand it, to extract the send_error_msg, I would have to get access to the `send_response` function, but that is only available while in the executor. So I can't move it outside the executer class?

**Comment by flying-sheep** [comment]
yes, therefore we should make it a method of the executor, i think

**Comment by flying-sheep** [comment]
maybe add the methods:

```r
request_is_silent = function(request) request$content$silent,
send_error_msg = function(request, msg) {
    if (!request_is_silent(request)) {
            send_response('stream', request, 'iopub', list(
                name = 'stderr',
                text = msg))
        }
    }
},
```

or so.

**Comment by janschulz** [comment]
IMO a lot of functions could be extracted if there would be a way to get the current request via an methods.

**Comment by janschulz** [comment]
Updated on top of master and repushed

**Pull request merged**